src/rdpc/analysesProcessor.ts

```diff
@@ -215,6 +310,117 @@ export const getAllMergedDonor = async ({
   return mergedDonors;
 };
 
+export const getAllMergedDonorWithSpecimens = async ({
+  analysesFetcher = fetchAnalysesWithSpecimens,
+  egoJwtManager,
+  studyId,
+  url,
+  config,
+  donorIds,
+}: {
+  studyId: string;
+  url: string;
+  egoJwtManager: EgoJwtManager;
+  donorIds?: string[];
+  config?: {
+    chunkSize?: number;
+    state?: StreamState;
+  };
+  analysesFetcher: typeof fetchAnalysesWithSpecimens;
+}): Promise<SpecimensByDonors> => {
+  let mergedDonors: SpecimensByDonors = {};
+
+  if (donorIds) {
+    for (const donorId of donorIds) {
+      logger.info(`streaming analyses with Specimens for donor ${donorId}`);
+      const stream = analysisStream_withSpeicmens({
+        studyId,
+        rdpcUrl: url,
+        egoJwtManager,
+        config,
+        analysesFetcher,
+        donorId,
+      });
+      for await (const page of stream) {
+        if (page.length === 0) {
+          logger.info(
+            `No sequencing experiment analyses with specimens fetched`
+          );
+        }
+        logger.info(
+          `Streaming ${page.length} of sequencing experiment analyses with specimens...`
+        );
+        const donorPerPage = aggregateSpecimensByDonorId(page);
+        mergeAllPagesSpecimensByDonorId(mergedDonors, donorPerPage);
+      }
+    }
+  } else {
+    const stream = analysisStream_withSpeicmens({
+      studyId,
+      rdpcUrl: url,
+      egoJwtManager,
+      config,
+      analysesFetcher,
+    });
+    for await (const page of stream) {
+      if (page.length === 0) {
+        logger.info(
+          `No sequencing experiment analyses with specimens for streaming`
+        );
+      }
+      logger.info(
+        `Streaming ${page.length} of sequencing experiment analyses with specimens...`
+      );
+      const donorPerPage = aggregateSpecimensByDonorId(page);
+      mergeAllPagesSpecimensByDonorId(mergedDonors, donorPerPage);
+    }
+  }
+  return mergedDonors;
+};
+
+export const mergeAllPagesSpecimensByDonorId = (
+  merged: SpecimensByDonors,
+  toMerge: SpecimensByDonors
+): SpecimensByDonors => {
+  Object.entries(toMerge).forEach(([donorId, specimens]) => {
+    const combined = merged[donorId]
+      ? mergedSpecimens(specimens, merged[donorId])
+      : specimens;
+    merged[donorId] = combined;
+  });
+  return merged;
```

Comment:
No uses of this method use the returned object, instead this method is modifying the contents of `merged`.

In the current implementation we can remove this, but we should consider finding a way to re-write `fetchAnalysesWithSpecimens` that doesn't require this method to modify the contents of its arguments. For instance, we could add all our specimens to a merged list, and then at the end of all the loops we can create a new object that has all the duplicates filtered out and return that object.

Reply:
added return object for testing purpose, forgot to remove it.